### Load functions needed for running R2 D2 model
source("R2_D2_model.R", echo = FALSE)


### Set your working directory
setwd("\\path_to_your_working_directory\\")


### Note: the below getPhenotypeXXX functions are for extracting gene expression from array data in BrainCloud and 4BrainR datasets.
### Users can generate expression files on their own without these functions as long as the output expression file conforms to
### the following 3-column format (header should not be changed):
### FID IID expression

# ***************** Preprocessing steps ******************** #
### Extract mRNA expression data (phenotype)for a specfic probe from the BrainCloud datasets
### Params:
###   - "gene", name of the gene of interest
###   - "probe.name", expression_file, probe_file
###   - "expr.dir", e.g., "BC-expression-GSE30272_res.txt" : stores the residuals of gene expression after regressing out age, gender factors
###   - "probe.dir", e.g., "probe-info-BC": stores the probe information for each gene
getPhenotypeBC("GSTM3", "HEEBO-054-HCC54J15", "BC-expression-GSE30272_res.txt", "probe-info-BC")

### Extract mRNA expression data (phenotype)for a specfic probe from the 4BrainR datasets
### This function is similar as the above one, but with different set of two input files from the 4BrainR dataset
###   - "4brainR-expression-CERE.txt", 4brainR-expression-FCTX.txt",'4brainR-expression-PONS.txt', or, "4brainR-expression-TCTX.txt"
###   - "probe-info-BC", probe information for each gene for the 4BrainR dataset
getPhenotypeCERE("MTHFR", "ILMN_1731434", "4brainR-expression-CERE.txt", "4BrainR-probe-info.txt")
getPhenotypeFCTX("MTHFR", "ILMN_1731434", "4brainR-expression-FCTX.txt", "4BrainR-probe-info.txt")
getPhenotypePONS("MTHFR", "ILMN_1731434", "4brainR-expression-PONS.txt", "4BrainR-probe-info.txt")
getPhenotypeTCTX("MTHFR", "ILMN_1731434", "4brainR-expression-TCTX.txt", "4BrainR-probe-info.txt")
# ***************** Preprocessing steps ends ******************** #

### Main Program to run the R2 D2 model
### Params:
###   - "geno", XXX.raw file which contains all genotypes. This is a standard output from "PLINK --recodeAD".
###   - "qasso", YYY.qassoc file which contains all associations. This is a standard output from PLINK. extensions ".raw" and ".qassoc" need not be entered
###   - "num", 1,2,3 or 4, indicating the the analysis of random combinations of 1-4 candidate iSNPs
###   - "save_interm", boolean, whether to save intermediate files such as rGG.
### Note: format of geno files is as follows where the first six columns are extra information for each individual:
###       FID IID PAT MAT SEX PHENOTYPE snp1_A snp2_G ...
###       1 1 0 0 1 1  2 0 ...
###       1 2 0 0 2 1  1 1 ...
### Note: the analysis of 4-iSNP models may require very long computation times for chromosome regions ###
### interest (ROI) containing a large number of SNPs!
### Output:
###   - file with suffix of "snps-m.txt", which listing all models ranked by the evaluating metrics, (default NRMSE)
###   - file with suffix of "snps-plot-m.txt", which is used for plotting the best model with familyPlot_matrix().
comTest_matrix(geno = "GSTM-BC", qassoc = "GSTM3_HEEBO-054-HCC54J15", 3, save_interm = T)


### Plotting the results
### Params:
###   - The first parameter is the file generated by the Main program with the suffix of "3snps-plot-m.txt", the "3" corresponds to number of index SNPs in selected model.
###   - The second parameter is the user-defined name of output PDF file.
###   - The last parameter is user-defined colors for plotting lines. For more color options, visit: http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf
### Output:
###   - A PDF file in which the best model is plotted.
familyPlot_matrix("GSTM3_HEEBO-054-HCC54J15-3snps-plot-m", "GSTM3-BC-3snps", color = c("red", "green", "black", "brown"))

### Plotting the results of analyses for 1,2,3 or 4 candidate iSNPs selected by the user.
### Params:
###   - "geno", XXX.raw file.
###   - "qassoc", XXX.qassoc file. ".raw" and ".qassoc" extensions need not be entered
###   - "snplist", user defined and comma-separated list of iSNPS
###   - "color", user-defined colors for plotting lines. For more color options, visit: http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf
### Output:
###   - A PDF file in which the user-chosen model is plotted.
familyPlot_matrix_snps(geno = "GSTM-BC", qassoc ="GSTM3_HEEBO-054-HCC54J15", snplist = c("rs1296955","rs7536162", "rs11101999", "rs17597934"), color = c("red", "green", "black", "brown"))

